File: backend/app/schemas/timeline.py
"""Timeline lattice and dyadic refinement schemas."""

from pydantic import BaseModel, Field
from typing import List, Dict, Any, Optional

class SubBlockItem(BaseModel):
    child_id: str
    parent_id: int
    frame_start: int
    frame_end_exclusive: int
    lower_bound: float
    upper_bound: float
    truth_value: str
    status: str

class BlockLatticeItem(BaseModel):
    block_id: int
    frame_start: int
    frame_end_exclusive: int
    time_start_ns: float
    time_end_ns: float
    lower_bound: float
    upper_bound: float
    truth_value: str  # 'TRUE', 'FALSE', 'UNKNOWN'
    status: str       # 'CERTIFIED_TRUE', 'CERTIFIED_FALSE', 'REFINED', 'EXACT', 'UNKNOWN'
    child_blocks: List[SubBlockItem] = Field(default_factory=list)
    exact_frames: int = 0
    refined_count: int = 0

class BlockRefineRequest(BaseModel):
    block_id: int
    trajectory_id: Optional[str] = "protein_ligand_traj.xtc"
    subdivision_factor: int = 2

class BlockRefineResponse(BaseModel):
    parent_block_id: int
    parent_bounds: Dict[str, float]
    child_blocks: List[SubBlockItem]
    monotonic_non_expansion_verified: bool
    summary: str

File: frontend/src/types/timeline.ts
/**
 * Timeline block lattice and dyadic refinement types.
 */

export interface SubBlockItem {
  child_id: string;
  parent_id: number;
  frame_start: number;
  frame_end_exclusive: number;
  lower_bound: number;
  upper_bound: number;
  truth_value: string;
  status: string;
}

export interface BlockLatticeItem {
  block_id: number;
  frame_start: number;
  frame_end_exclusive: number;
  time_start_ns: number;
  time_end_ns: number;
  lower_bound: number;
  upper_bound: number;
  truth_value: 'TRUE' | 'FALSE' | 'UNKNOWN' | string;
  status: 'CERTIFIED_TRUE' | 'CERTIFIED_FALSE' | 'REFINED' | 'EXACT' | 'UNKNOWN' | string;
  child_blocks: SubBlockItem[];
  exact_frames: number;
  refined_count: number;
}

export interface BlockRefineResponse {
  parent_block_id: number;
  parent_bounds: { L: number; U: number };
  child_blocks: SubBlockItem[];
  monotonic_non_expansion_verified: boolean;
  summary: string;
}
